src/commands/deck.ts

```diff
@@ -340,12 +368,71 @@ export class DeckCommand extends Command {
 		const outFile = typedDeckToYdk(deck);
 
 		const content = await this.generateProfile(deck, resultLanguage, !isStacked, outUrl);
+
+		// prepare interaction button for FTP upload
+		const row = new ActionRowBuilder<ButtonBuilder>().addComponents(
+			new ButtonBuilder().setCustomId("ftp").setLabel("Upload to YGOPRODeck").setStyle(ButtonStyle.Primary)
+		);
+
+		// a string is interpreted as a path, to upload it as a file we need a Buffer
+		// we might also want it again for the FTP upload
+		const deckBuffer = Buffer.from(outFile, "utf-8");
+
+		// save to re-use when we later edit the message
+		const embeds = addNotice(content);
+		const attachment = new AttachmentBuilder(deckBuffer).setName("deck.ydk");
+
 		const end = Date.now();
-		await interaction.editReply({
-			embeds: addNotice(content),
-			// a string is interpreted as a path, to upload it as a file we need a Buffer
-			files: [new AttachmentBuilder(Buffer.from(outFile, "utf-8")).setName("deck.ydk")]
+		const response = await interaction.editReply({
+			embeds,
+			files: [attachment],
+			components: [row]
 		});
+
+		const filter = (i: ButtonInteraction): boolean => {
+			// only allow the OP to click
+			return i.user.id === interaction.user.id;
+		};
+
+		// TODO: determine ideal timeout length.
+		// we don't await this promise, we set up the callback and then let the method complete
+		response
+			.awaitMessageComponent({ filter, componentType: ComponentType.Button, time: 60000 })
+			.then(async i => {
+				await i.deferReply({ ephemeral: !isPublic });
+
+				// upload deck with FTP
+				const filename = `${i.id}.ydk`; // use interaction ID for unique filename
+				await this.upload(filename, deckBuffer);
+
+				// disable original button
+				// prepare row to disable button on original message
+				useLocale(resultLanguage);
+				const disabledRow = new ActionRowBuilder<ButtonBuilder>().addComponents(
+					new ButtonBuilder()
+						.setCustomId("ftp-disabled")
+						.setLabel(t`Upload Complete`)
+						.setStyle(ButtonStyle.Success)
+						.setDisabled(true)
+				);
+				await interaction.editReply({ embeds, files: [attachment], components: [disabledRow] });
+
+				// reply in affirmation
+				await i.editReply({
+					content: t`Deck successfully uploaded to <https://ygoprodeck.com/deckbuilder/?u=https://ygoprodeck.com/discord-decks/${filename}>!`,
+					components: []
+				});
+			})
+			.catch(async (err: Error) => {
+				// a rejection can mean the timeout was reached without a response
+				// otherwise, though, we want to treat it as a normal error
+				if (err.name !== "Error [InteractionCollectorError]") {
+					throw err;
+				}
```

Comment:
I see what you're gunning for here. A shared error-handling stack for this, like already happens for ChatInputCommandInteraction. Having the logic for the component interactions next to the associated command is convenient, since listening for a MessageComponentInteraction would separate the button from what it does quite a bit.

Reply:
Yeah. One way to make error handling more manageable would be to have a Component system parallel to the Command system, but if they're separated like that it would be difficult, or at least tedious, to pass significant amounts of information from the Command Closure to the Component. I'd be happy just replacing `throw err` here with `foo(err)`, where foo is an important error-handling function, I'm just not sure how best to integrate the details of *that* function into our existing logging and metrics systems.

Reply:
If we do have an external function, note that it's difficult to get the button interaction context in the `.catch()` callback (and so potentially pass it to a logging function). It's only given to `.then()`.

Reply:
I'll sleep on the greater architecture and see what I can cook up...